addons/stock_landed_costs/models/stock_landed_costs.py

```diff
@@ -0,0 +1,273 @@
+                cost._create_accounting_entries(line, move, qty_out)
+            cost.write({'state': 'done', 'account_move_id': move.id})
+            cost.account_move_id.post()
+        return True
+
+    @api.multi
+    def button_cancel(self):
+        for cost in self:
+            if cost.state == 'done':
+                raise UserError(_('Validated landed costs cannot be cancelled, '
+                                'but you could create negative landed costs to reverse them'))
+            cost.write({'state': 'cancel'})
+        return True
+
+    @api.multi
+    def unlink(self):
+        # cancel or raise first
+        self.button_cancel()
+        return super(StockLandedCost, self).unlink()
+
+    @api.multi
+    def compute_landed_cost(self):
+        StockValuationLine = self.env['stock.valuation.adjustment.lines']
+        total_qty = total_cost = total_weight = total_volume = total_line = 0.0
+        digits = dp.get_precision('Product Price')(self.env.cr)
+        self.valuation_adjustment_line_ids.unlink()
+        for cost in self.filtered(lambda cost: cost.picking_ids):
+            for v in cost.get_valuation_lines():
+                for line in cost.cost_line_ids:
+                    v.update({'cost_id': cost.id, 'cost_line_id': line.id})
+                    StockValuationLine.create(v)
+                total_qty += v.get('quantity', 0.0)
+                total_cost += v.get('former_cost', 0.0)
+                total_weight += v.get('weight', 0.0)
+                total_volume += v.get('volume', 0.0)
+                total_line += 1
+            for line in cost.cost_line_ids:
+                value_split = 0.0
+                for valuation in cost.valuation_adjustment_line_ids.filtered(lambda l: l.cost_line_id.id == line.id):
+                    value = 0.0
+                    if line.split_method == 'by_quantity' and total_qty:
+                        value = valuation.quantity * (line.price_unit / total_qty)
+                    elif line.split_method == 'by_weight' and total_weight:
+                        value = valuation.weight * (line.price_unit / total_weight)
+                    elif line.split_method == 'by_volume' and total_volume:
+                        value = valuation.volume * (line.price_unit / total_volume)
+                    elif line.split_method == 'by_current_cost_price' and total_cost:
+                        value = valuation.former_cost * (line.price_unit / total_cost)
+                    else:
+                        value = (line.price_unit / total_line)
+
+                    if digits:
+                        value = float_round(value, precision_digits=digits[1], rounding_method='UP')
+                        fnc = min if line.price_unit > 0 else max
+                        value = fnc(value, line.price_unit - value_split)
+                        value_split += value
+
+                    valuation.additional_landed_cost += value
```

Comment:
Be aware of using Active Record Pattern, as each assignment will trigger write on database .